src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1375,61 +1376,102 @@ uct_dc_mlx5_iface_dci_do_rand_pending_tx(ucs_arbiter_t *arbiter,
+    if (ucs_unlikely(req->func == uct_dc_mlx5_iface_fc_grant)) {
+        /* User callback should not be called for FC messages. Just return
+         * pending request memory to the pool */
+        freq = ucs_derived_of(req, uct_rc_pending_req_t);
+        ucs_mpool_put(freq);
+        return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
+    }
+
+    /* Non-internal request was found */
+    return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
+}
+
+static ucs_arbiter_cb_result_t
+uct_dc_mlx5_ep_arbiter_purge_cb(ucs_arbiter_t *arbiter, ucs_arbiter_group_t *group,
+                                ucs_arbiter_elem_t *elem, void *arg)
+{
+    uct_purge_cb_args_t *cb_args = arg;
+    void **priv_args             = cb_args->arg;
+    uct_dc_mlx5_ep_t *ep         = priv_args[0];
+    uct_pending_req_t *req       = ucs_container_of(elem, uct_pending_req_t,
+                                                    priv);
+    ucs_arbiter_cb_result_t result;
+
+    result = uct_dc_mlx5_ep_arbiter_purge_internal_cb(arbiter, group, elem,
+                                                      arg);
+    if (ucs_likely(result == UCS_ARBITER_CB_RESULT_RESCHED_GROUP)){
         if (cb_args->cb != NULL) {
             cb_args->cb(req, priv_args[1]);
         } else {
             ucs_debug("ep=%p cancelling user pending request %p", ep, req);
         }
-    } else {
-        /* User callback should not be called for FC messages.
-         * Just return pending request memory to the pool */
-        freq = ucs_derived_of(req, uct_rc_pending_req_t);
-        ucs_mpool_put(freq);
+        return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
     }
 
-    return UCS_ARBITER_CB_RESULT_REMOVE_ELEM;
+    return result;
 }
 
-void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep, uct_pending_purge_callback_t cb, void *arg)
+/* Return value indicates whether purging should continue or not */
+int uct_dc_mlx5_ep_pending_purge_internal(uct_dc_mlx5_ep_t *ep,
+                                          uct_dc_mlx5_iface_t *iface,
+                                          ucs_arbiter_callback_t arbiter_cb,
+                                          uct_pending_purge_callback_t cb,
+                                          void *arg)
 {
-    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
-    uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
-    void *priv_args[2]         = {ep, arg};
-    uct_purge_cb_args_t args   = {cb, priv_args};
+    void *priv_args[2]       = {ep, arg};
+    uct_purge_cb_args_t args = {cb, priv_args};
     ucs_arbiter_t *waitq;
     ucs_arbiter_group_t *group;
     uint8_t pool_index;
-    uint8_t dci;
 
     if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
         ucs_arbiter_group_purge(uct_dc_mlx5_iface_tx_waitq(iface),
                                 uct_dc_mlx5_ep_rand_arb_group(iface, ep),
-                                uct_dc_mlx5_ep_arbiter_purge_cb, &args);
-        return;
+                                arbiter_cb, &args);
```

Comment:
looks like with rand scheme it should segfault

Reply:
why? it should work the same way as previously